### ISSUE TITLE
autosnippets: "alpha" -> "\alpha" but not "\alpha" -> "\\alpha"

### ISSUE THREAD

**Issue by IndianBoy42** [L3MON4D3/LuaSnip#150]
I am trying to make an autosnippet to automatically add backslashes for some common latex words. If i make the trigger just `"alpha"` then it would trigger even when I already added a backslash. Also a related problem is that if i have such a snippet for beta and eta. then when I actually write beta it may become `b\eta`

I tried: `[^%\]alpha` but this doesn't trigger except at the beginning of the line, or as part of another word. It seem it doesn't include the preceeding whitespace in the pattern search?

Basically I need some kind of better detection of word boundaries or preceeding whitespace in regex triggers

**Comment by kmarius** [comment]
I don't think this can be easily done with lua patterns. Try this condition function:
```
local line_to_cursor = require("luasnip.util.util").get_current_line_to_cursor

local function no_backslash()
	return not string.match(line_to_cursor(), "\\%w+$")
end

```

**Comment by L3MON4D3** [comment]
> Also a related problem is that if i have such a snippet for beta and eta. then when I actually write beta it may become `b\eta`

Could you try defining the snippet for `beta` before `eta`? Luasnip tries to match the snippets in the order they were defined, so this way `eta` could have precedence.

**Comment by L3MON4D3** [comment]
> I tried: `[^%\]alpha` but this doesn't trigger except at the beginning of the line, or as part of another word. It seem it doesn't include the preceeding whitespace in the pattern search?

That does sound like a bug, I'll be able to look into it tomorrow

**Comment by IndianBoy42** [comment]
> > Also a related problem is that if i have such a snippet for beta and eta. then when I actually write beta it may become `b\eta`
> 
> Could you try defining the snippet for `beta` before `eta`? Luasnip tries to match the snippets in the order they were defined, so this way `beta` should have precedence.

Hmm, beta is defined before eta. The behavior i mentioned here is if i define the trigger as "[^%\]beta". So with trying to ignore the backslash. Again i think it's because the matcher doesn't include preceding whitespace, so the first `[]` matches the b, and then it expands the eta. Instead of the first `[]` matching the space and then expanding the beta

**Comment by L3MON4D3** [comment]
Oh, okay then the issue should be fixed first :D
The matching for pattern-triggers basically appends a '$' to the pattern and then checks whether it matches the line up to the cursor (see [here](https://github.com/L3MON4D3/LuaSnip/blob/1e79af9cfe74bcce13e8646cd7809d915f218504/lua/luasnip/nodes/snippet.lua#L345)), so I'm not sure what exactly is going wrong. Have you tried the trigger without escaping the `\`?

**Comment by L3MON4D3** [comment]
Okay, I played around with it a bit, `[^\]beta` only expands if there is another char in front of it, `[^\]?beta` doesn't work because lua tries to make the pattern match, and so matches the `beta` in `\beta`.
Everything works as expected, afaict.
What you could do is:
```lua
-- https://gist.github.com/tylerneylon/81333721109155b2d244
local function copy3(obj, seen)
	-- Handle non-tables and previously-seen tables.
	if type(obj) ~= "table" then
		return obj
	end
	if seen and seen[obj] then
		return seen[obj]
	end

	-- New table; mark it as seen an copy recursively.
	local s = seen or {}
	local res = {}
	s[obj] = res
	for k, v in next, obj do
		res[copy3(k, s)] = copy3(v, s)
	end
	return setmetatable(res, getmetatable(obj))
end

-- only guessing here :D
local beta_nodes = {t"\\beta"}

ls.autosnippets = {
	all = {
		-- match in line.
		-- copy to not override edges in second constructor.
		-- "end beta|" would not match with wordTrig because the match is " beta"
		-- and the char before the trigger "e", which is in [%w_].
		s({trig = "[^\\]beta", regTrig = true, wordTrig=false}, copy3(beta_snip)),
		-- match at beginning of line.
		s({trig = "^beta", regTrig = true}, copy3(beta_snip))
	}
}
```

This would keep duplicate code to a minimum, but tbh, a condition-function would
be nicer, and probably faster, which is a good bonus with autosnippets.
